### ISSUE TITLE
[LLVM19] library packages are no longer multi-arch capable (installation of eg. i386 and amd64 in parallel is impossible)

### ISSUE THREAD

**Issue by TheRealCuran** [llvm/llvm-project#82779]
The newest batch of LLVM 19 packages from apt.llvm.org (1:19~++20240222101030+ae3e14276b71-1~exp1~20240222221201.1911) install binary files like `libLLVM` no longer into their multi-arch directories, but into `/usr/lib` directly:

```
$ dpkg-deb -c libllvm19_1%3a19~++20240222101030+ae3e14276b71-1~exp1~20240222221201.1911_amd64.deb 
drwxr-xr-x root/root         0 2024-02-22 23:12 ./
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/lib/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/lib/llvm-19/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/lib/llvm-19/lib/
-rw-r--r-- root/root 124842264 2024-02-22 23:12 ./usr/lib/llvm-19/lib/libLLVM.so.1
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/lib/x86_64-linux-gnu/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/doc/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/doc/libllvm19/
-rw-r--r-- root/root     36373 2024-02-22 23:12 ./usr/share/doc/libllvm19/changelog.Debian.gz
-rw-r--r-- root/root      6897 2024-02-22 23:09 ./usr/share/doc/libllvm19/copyright
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/lintian/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/lintian/overrides/
-rw-r--r-- root/root       122 2024-02-22 23:12 ./usr/share/lintian/overrides/libllvm19
lrwxrwxrwx root/root         0 2024-02-22 23:12 ./usr/lib/llvm-19/lib/libLLVM.so.19.0 -> libLLVM.so.1
lrwxrwxrwx root/root         0 2024-02-22 23:12 ./usr/lib/x86_64-linux-gnu/libLLVM-19.so -> libLLVM-19.so.1
$ dpkg-deb -c libllvm19_1%3a19~++20240222101030+ae3e14276b71-1~exp1~20240222221201.1911_i386.deb 
drwxr-xr-x root/root         0 2024-02-22 23:12 ./
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/lib/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/lib/i386-linux-gnu/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/lib/llvm-19/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/lib/llvm-19/lib/
-rw-r--r-- root/root 140601144 2024-02-22 23:12 ./usr/lib/llvm-19/lib/libLLVM.so.1
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/doc/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/doc/libllvm19/
-rw-r--r-- root/root     36373 2024-02-22 23:12 ./usr/share/doc/libllvm19/changelog.Debian.gz
-rw-r--r-- root/root      6897 2024-02-22 23:09 ./usr/share/doc/libllvm19/copyright
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/lintian/
drwxr-xr-x root/root         0 2024-02-22 23:12 ./usr/share/lintian/overrides/
-rw-r--r-- root/root       122 2024-02-22 23:12 ./usr/share/lintian/overrides/libllvm19
lrwxrwxrwx root/root         0 2024-02-22 23:12 ./usr/lib/i386-linux-gnu/libLLVM-19.so -> libLLVM-19.so.1
lrwxrwxrwx root/root         0 2024-02-22 23:12 ./usr/lib/llvm-19/lib/libLLVM.so.19.0 -> libLLVM.so.1
```
On top of that the names and symlinks are really messed up too (`libLLVM.so.1`) can only be in one package, if the next LLVM 20 packages try to ship this too, then you can't have multiple LLVM versions in parallel. Something that is quite essential on how breakage-happy LLVM is across major versions. And also the `libLLVM-19.so` symlink just leads nowhere.

This makes the current packages unusable on a multi-arch setup as for example needed with 32 bit applications, requiring eg. OpenGL through Mesa, as I can't install `libllvm19:amd64` and `libllvm19:i386` in parallel.

Please restore the previous behaviour of having the binary arch-dependant files in their respective multi-arch directories like `/usr/lib/x86_64-linux-gnu` and have them named in a way, that makes having multiple LLVM versions installed and usable in parallel. Settling on something like `libLLVM.so.19.0` instead of `libLLVM-19.so.1` is perfectly fine, just no `libLLVM.so.1` please.

**Comment by LW-archlinux** [comment]
Not a duplicate but has similariites with https://github.com/llvm/llvm-project/issues/82647 

Also check https://github.com/llvm/llvm-project/pull/82660

**Comment by TheRealCuran** [comment]
Just to make sure: this is not a simple fix in the `.install` files. You need to make sure that e.g. the generated CMake files also contain the correct paths and names.

Locally I have "fixed" all of this and been able to build a new round of the Mesa stack including the `spirv-llvm-translator`.

**Comment by TheRealCuran** [comment]
The conflicting `libLLVM.so.1` and broken symlinks are still present in `1:19~++20240228111504+fd336c33b6c9-1~exp1~20240228111638.1922`

**Comment by sylvestre** [comment]
ok :/
it will take me sometime to bring back the multi arch capability after these changes :(

**Comment by TheRealCuran** [comment]
Thanks for the update. Then I'll just have to keep "fixing" (ie. moving files and editing eg. `LLVMExports-relwithdebinfo.cmake`) locally. Not nice, but it'll work for a while.

**Comment by TheRealCuran** [comment]
This particular issue is fixed for me with the current LLVM 19 packages from Debian and also the LLVM 20 packages from apt.llvm.org, therefore I am closing this issue.

**Comment by serhii-nakon** [comment]
Currently no i386 packages in repositories at all...

**Comment by sylvestre** [comment]
@serhii-nakon for which distro version? thanks

**Comment by sylvestre** [comment]
for example, Debian unstable has it:
https://apt.llvm.org/unstable/pool/main/l/llvm-toolchain-snapshot/libclang-common-20-dev_20~++20241022100635+d98519715617-1~exp1~20241022220750.2418_i386.deb

**Comment by serhii-nakon** [comment]
@sylvestre I use Debian Bookworm - here no packages for i386 at all

**Comment by sylvestre** [comment]
@serhii-nakon could you please a new issue? It is unrelated to this one.
thanks

**Comment by serhii-nakon** [comment]
@sylvestre Ok

**Comment by sylvestre** [comment]
thanks :)

**Comment by serhii-nakon** [comment]
@sylvestre https://github.com/llvm/llvm-project/issues/113438
